### ISSUE TITLE
Feature request i2c serial eeproms

### ISSUE THREAD

**Issue by orphanfour** [NaejEL/flipperzero-i2ctools#8]
Would it be possible to implement serial eeprom read/write over i2c? This would be very useful for reverse engineering. I have constructed a 3.3v to 5v bridge using an ISO1541 ic and can access data on a xicor x24c04 serial eeprom but reading/writing the individual bytes one at a time is somewhat useless. If you could add the option to dump and write i2c eeproms to and from a file it would expand the usefulness of the tool a great deal. If you need hardware to test it on i can offer mine. Thanks for everything you have done so far.

**Comment by NaejEL** [comment]
Good idea. Flipper Zero have a tool named "SPI Mem Manager" but no "I2C Mem Manager". I will give it a try. As I understand the x24c04 datasheet, the eeprom is composed of 512 "words" of 8bits.

**Comment by orphanfour** [comment]
![RIMG0883](https://github.com/NaejEL/flipperzero-i2ctools/assets/69114558/fb1dd0b6-cb3f-4362-b082-a8670aa76b2d)
Heres a picture of my adaptor board with the ISO1541 isolator/level converter between the flipper gpio and x24c04 It does work but can only read two bytes at a time with your existing i2c app.

**Comment by orphanfour** [comment]
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include "flipper.h" // Include Flipper SDK header

#define EEPROM_I2C_ADDRESS 0x50
#define BUFFER_SIZE 256

static uint8_t buffer[BUFFER_SIZE];

void read_eeprom(uint16_t start_address, uint16_t length) {
    if (length > BUFFER_SIZE) {
        printf("Error: Length exceeds buffer size.\n");
        return;
    }
    for (uint16_t i = 0; i < length; i++) {
        buffer[i] = furi_hal_i2c_read_byte(EEPROM_I2C_ADDRESS, start_address + i);
    }
    printf("Data read successfully.\n");
}

void write_eeprom(uint16_t start_address, uint16_t length) {
    if (length > BUFFER_SIZE) {
        printf("Error: Length exceeds buffer size.\n");
        return;
    }
    for (uint16_t i = 0; i < length; i++) {
        furi_hal_i2c_write_byte(EEPROM_I2C_ADDRESS, start_address + i, buffer[i]);
    }
    printf("Data written successfully.\n");
}

void save_to_file(const char* filename) {
    FILE* file = fopen(filename, "wb");
    if (!file) {
        printf("Error: Unable to open file for writing.\n");
        return;
    }
    fwrite(buffer, sizeof(uint8_t), BUFFER_SIZE, file);
    fclose(file);
    printf("Data saved to %s.\n", filename);
}

void load_from_file(const char* filename) {
    FILE* file = fopen(filename, "rb");
    if (!file) {
        printf("Error: Unable to open file for reading.\n");
        return;
    }
    fread(buffer, sizeof(uint8_t), BUFFER_SIZE, file);
    fclose(file);
    printf("Data loaded from %s.\n", filename);
}

void calculate_checksum() {
    uint16_t checksum = 0;
    for (uint16_t i = 0; i < BUFFER_SIZE; i++) {
        checksum += buffer[i];
    }
    printf("Checksum: 0x%04X\n", checksum);
}

void display_menu() {
    printf("\nI2C EEPROM Tool\n");
    printf("1. Read EEPROM\n");
    printf("2. Write EEPROM\n");
    printf("3. Save Buffer to File\n");
    printf("4. Load Buffer from File\n");
    printf("5. Calculate Checksum\n");
    printf("6. Exit\n");
}

int main() {
    int choice;
    uint16_t start_address, length;
    char filename[256];

    while (1) {
        display_menu();
        printf("Select an option: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter start address (hex): ");
                scanf("%hx", &start_address);
                printf("Enter length: ");
                scanf("%hu", &length);
                read_eeprom(start_address, length);
                break;

            case 2:
                printf("Enter start address (hex): ");
                scanf("%hx", &start_address);
                printf("Enter length: ");
                scanf("%hu", &length);
                write_eeprom(start_address, length);
                break;

            case 3:
                printf("Enter filename: ");
                scanf("%s", filename);
                save_to_file(filename);
                break;

            case 4:
                printf("Enter filename: ");
                scanf("%s", filename);
                load_from_file(filename);
                break;

            case 5:
                calculate_checksum();
                break;

            case 6:
                printf("Exiting...\n");
                return 0;

            default:
                printf("Invalid choice. Please try again.\n");
                break;
        }
    }

    return 0;
}
